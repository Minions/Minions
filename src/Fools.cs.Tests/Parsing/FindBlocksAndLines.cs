// FindBlocksAndLines.cs
// 
// Copyright 2012 The Minions Project (http:/github.com/Minions).
// All rights reserved. Usage as permitted by the LICENSE.txt file for this project.

using ApprovalTests.Reporters;
using Fools.cs.Tests.Support;
using NUnit.Framework;

namespace Fools.cs.Tests.Parsing
{
    [TestFixture, UseReporter(typeof (DiffReporter))]
    public class FindBlocksAndLines
    {
        [Test]
        public void should_find_nothing_in_an_empty_file()
        {
            string.Empty.find_blocks()
                .should_fail();
        }

        [Test]
        public void file_with_a_single_simple_block_should_parse()
        {
            @"
def foo():
	a = b
	

".find_blocks()
                .should_parse_correctly();
        }

        [Test]
        public void empty_blocks_should_be_explicitly_stated()
        {
            "def foo():\r\n\tpass".find_blocks()
                .should_parse_correctly();
        }

        [Test]
        public void blocks_without_bodies_should_fail()
        {
            "def foo():\r\na=b".find_blocks()
                .should_fail();
        }

        [Test]
        public void blocks_with_only_empty_lines_should_fail()
        {
            "def foo():\r\n\t\r\na=b".find_blocks()
                .should_fail();
        }

        [Test]
        public void should_reject_illegal_indentation_with_good_error_message()
        {
            "def foo():\r\n\tb = c\r\n\t\ta = b".find_blocks()
                .should_fail();
        }

        [Test]
        public void spaces_at_start_of_a_line_should_be_illegal()
        {
            "def foo():\r\n\t b = c".find_blocks()
                .should_fail();
        }

        [Test]
        public void spaces_in_the_indentation_should_be_illegal()
        {
            "def foo():\r\n\tif bar:\r\n\t \tb = c".find_blocks()
                .should_fail();
        }

        [Test]
        public void undefined_block_type_should_be_illegal()
        {
            "frazzle:\r\n\tpass".find_blocks()
                .should_fail();
        }

        [Test]
        public void only_declaration_blocks_should_be_allowed_at_the_top_level()
        {
            "if bar:\r\n\tpass".find_blocks()
                .should_fail();
        }

        [Test]
        public void no_statements_should_be_allowed_at_the_top_level()
        {
            "a=b".find_blocks()
                .should_fail();
        }
    }
}
