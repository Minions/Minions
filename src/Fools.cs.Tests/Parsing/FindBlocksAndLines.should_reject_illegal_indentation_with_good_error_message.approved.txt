Should fail

/** Declarations **/
[]

/** Errors **/
[
  {
    "$type": "Fools.cs.ParseToAst.ErrorReport, Fools.cs",
    "file_name": "fake file name.fool",
    "line": 3,
    "error_type": "Inconsistent indentation",
    "expected": 1,
    "actual": 2,
    "snippet": "\t\ta = b",
    "suggestion": "Doublecheck the amount of indentation on this line and on the previous line. The most common cause of this error is accidentally de-indenting a line, which will cause the next indented line to fail."
  },
  {
    "$type": "Fools.cs.ParseToAst.ErrorReport, Fools.cs",
    "file_name": "fake file name.fool",
    "line": 3,
    "error_type": "Unrecognized declaration",
    "expected": null,
    "actual": null,
    "snippet": "\t\ta = b",
    "suggestion": "I could not figure out this declaration. Is it a non-declaration statement? Recall that the top level of a Fools file is only allowed to contain declarations."
  },
  {
    "$type": "Fools.cs.ParseToAst.ErrorReport, Fools.cs",
    "file_name": "fake file name.fool",
    "line": 1,
    "error_type": "Failed to parse 'program'.",
    "expected": null,
    "actual": null,
    "snippet": null,
    "suggestion": "Hopefully I also gave you some more specific error messages. Try fixing those first."
  }
]
